sky/serve/replica_managers.py

```diff
@@ -884,22 +883,22 @@ def _refresh_process_pool(self) -> None:
                     logger.info(f'Replica {replica_id} removed from the '
                                 f'replica table {removal_reason}.')
 
-        # Clean old version
-        replica_infos = serve_state.get_replica_infos(self._service_name)
-        current_least_recent_version = min([
-            info.version for info in replica_infos
-        ]) if replica_infos else self.least_recent_version
-        if self.least_recent_version < current_least_recent_version:
-            for version in range(self.least_recent_version,
-                                 current_least_recent_version):
-                task_yaml = serve_utils.generate_task_yaml_file_name(
-                    self._service_name, version)
-                # Delete old version metadata.
-                serve_state.delete_version(self._service_name, version)
-                # Delete storage buckets of older versions.
-                service.cleanup_storage(task_yaml)
-            # newest version will be cleaned in serve down
-            self.least_recent_version = current_least_recent_version
+                    removed_version = info.version
+                    replica_infos = serve_state.get_replica_infos(
+                        self._service_name)
+                    no_replica_of_removed_version = all([
+                        info.version != removed_version
+                        for info in replica_infos
+                    ])
+                    if (no_replica_of_removed_version and
+                            removed_version != latest_version):
+                        task_yaml = serve_utils.generate_task_yaml_file_name(
+                            self._service_name, removed_version)
+                        # Delete old version metadata.
+                        serve_state.delete_version(self._service_name,
+                                                   removed_version)
+                        # Delete storage buckets of older versions.
+                        service.cleanup_storage(task_yaml)
```

Comment:
This is not equivalent to the original version handling. We should only remove the versions smaller than the least_recent_version. 

Original:
in db: 2 3 4 5
replicas: 3 5
we remove: 2

Now:
in db: 2 3 4 5
replica: 3 5
we remove: 2 4

Reply:
Did you mean version numbers in the example? I assumed that replicas of older versions would always be removed before newer versions, and that only the latest version wcould have replicas launched. It seems that is not true?